Add typing annotations to pattern_matcher.py

aorenste opened the pull request:
Turn on `mypy: disallow-untyped-defs` in pattern_matcher.py and fix the fallout.

There are still a bunch of `type: ignore` annotations which should eventually be ironed out.

In the processs found a bug: #127457 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127459
* __->__ #127458
* #127457



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

aorenste committed:
typing pattern_matcher.py

[ghstack-poisoned]

aorenste committed:
Update on "Add typing annotations to pattern_matcher.py"

Turn on `mypy: disallow-untyped-defs` in pattern_matcher.py and fix the fallout.

There are still a bunch of `type: ignore` annotations which should eventually be ironed out.

In the processs found a bug: #127457 




[ghstack-poisoned]

aorenste committed:
Update on "Add typing annotations to pattern_matcher.py"

Turn on `mypy: disallow-untyped-defs` in pattern_matcher.py and fix the fallout.

There are still a bunch of `type: ignore` annotations which should eventually be ironed out.

In the processs found a bug: #127457 




[ghstack-poisoned]

Skylion007 commented on torch/fx/node.py:
```diff
@@ -421,7 +422,7 @@ def insert_arg(self, idx : int, arg : Argument) -> None:
         self._args = args_left + (arg,) + args_right
 
-        _new_input_nodes = {}
+        _new_input_nodes: Dict[Node, None] = {}
```
Should the value be None or an Optional?

aorenste replied:
I think it can never have a value of anything other than none, can it? The only thing that mutates that dict is the `setdefault` on the following line - which will set the value to `None` (it's really using the dict as a set of keys)

Skylion007 commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -641,54 +726,60 @@ def pattern_eq(self, other: Any) -> bool:
-        assert self.outputs[0] and hasattr(self.outputs[0], "fns")
-        return self.outputs[0].fns
+    def fns(self) -> Union[Callable[..., Any], str, Sequence[Any]]:
+        output = typing.cast(_TargetExpr, self.outputs[0])
```
Could these protocol typing casts be settled with assert isinstance calls instead?

aorenste replied:
We already verify output[0] on line 733 above during __init__ - so another assert here would be redundant. I'll add a comment to that effect.

Skylion007 commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -59,8 +64,33 @@
+        ...
+
+
+T = TypeVar("T")
```
Does this TypeVar need to be be public? Or should it be prefixed with an underscore?

aorenste replied:
It's not public public - it's under _inductor so it's not in the public API.  On the other hand you're correct that it doesn't need to be visible outside this file. It seems like we generally have a lot of classes that fall into that category where as long as it's not in the public API we don't put the underscore.

Skylion007 reviewed (approved):
Some nits but definitely better typing than what exists now.

aorenste commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -320,21 +380,27 @@ class _TargetExpr(PatternExpr):
+        for fn in fns:
             if isinstance(fn, torch._ops.OpOverloadPacket):
-                fns.extend([getattr(fn, overload) for overload in fn.overloads()])
+                fns.extend(getattr(fn, overload) for overload in fn.overloads())
```
Still extend. Append would add a single entry. Since extend() is happy to take an iterator all this change does it make it not create an unnecessary temporary list.

aorenste committed:
Update on "Add typing annotations to pattern_matcher.py"

Turn on `mypy: disallow-untyped-defs` in pattern_matcher.py and fix the fallout.

There are still a bunch of `type: ignore` annotations which should eventually be ironed out.

In the processs found a bug: #127457 




[ghstack-poisoned]
